py/ringbuf.c

```diff
@@ -75,14 +75,8 @@ int ringbuf_put16(ringbuf_t *r, uint16_t v) {
     return 0;
 }
 
-// Returns:
-//    0: Success
-//   -1: Not enough data available to complete read (try again later)
-//   -2: Requested read is larger than buffer - will never succeed
-int ringbuf_get_bytes(ringbuf_t *r, uint8_t *data, size_t data_len) {
-    if (ringbuf_avail(r) < data_len) {
-        return (r->size <= data_len) ? -2 : -1;
-    }
+static inline MP_ALWAYSINLINE void _ringbuf_memcpy_get(ringbuf_t *r, uint8_t *data, size_t data_len) {
```

Comment:
I think it would be better to put these memcpy functions in `py/ringbuf.h` (and remove the leading underscore), so they can be public functions.

Then, the `RingIO` implementation can go in a separate file like `py/objringio.c`.

And, does this really need `MP_ALWAYSINLINE`?  If code size is unchanged without that attribute then I'd rather leave it out.

Reply:
Then we can see how the CI coverage looks with these changes, it may improve.

Reply:
I was wary of making them public because without adding a length check in the function calling them you can get a buffer overrun failure.

Without MP_ALWAYSINLINE it was bigger on rp2. Even with both `inline MP_ALWAYSINLINE` rp2 size increases by 4 compared to literally copying the contents into the target function, so something about the inline settings on the compiler there is tricky perhaps.

Reply:
My main motivation here is to put the implementation of `RingIO` in a separate file, because it really doesn't belong in the low-level ringbuf library code.

Can we put those helpers in `ringbuf.h` but named appropriately, eg `ringbuf_memcpy_get_internal`?  And a comment about why its "internal".

You can leave the `MP_ALWAYSINLINE` if it makes rp2 smaller.

Reply:
After the refactor I tested with/without `MP_ALWAYSINLINE` and get no change. Might also be realted to a different computer being used. Regardless, I don't think it's adding value anymore so have removed it.